src/shady/passes/reconvergence_heuristics.c

```diff
@@ -0,0 +1,473 @@
+    assert(parent);
+
+    while (parent) {
+        if (entries_count_list(parent->cf_nodes) != 1)
+            return false;
+
+        if (read_list(CFNode*, parent->cf_nodes)[0]->node == entry)
+            return true;
+
+        parent = parent->parent;
+    }
+
+    return false;
+}
+
+//TODO: This is massively inefficient.
+static void gather_exiting_nodes(LoopTree* lt, const CFNode* entry, const CFNode* block, struct List* exiting_nodes) {
+    if (!in_loop(lt, entry->node, block->node)) {
+        append_list(CFNode*, exiting_nodes, block);
+        return;
+    }
+
+    for (size_t i = 0; i < entries_count_list(block->dominates); i++) {
+        const CFNode* target = read_list(CFNode*, block->dominates)[i];
+        gather_exiting_nodes(lt, entry, target, exiting_nodes);
+    }
+}
+
+static const Node* process_node(Context* ctx, const Node* node) {
+    if (node == NULL) return NULL;
+
+    Rewriter * rewriter = &ctx->rewriter;
+    IrArena * arena = rewriter->dst_arena;
+
+    const Node* already_done = search_processed(&ctx->rewriter, node);
+    if (already_done)
+        return already_done;
+
+    if (is_function(node)) {
+        ctx->current_fn = node;
+    }
+
+    if (!ctx->current_fn || !lookup_annotation(ctx->current_fn, "Restructure")) {
+        return recreate_node_identity(&ctx->rewriter, node);
+    }
+
+    if (is_function(node)) {
+        if (ctx->fwd_scope) {
+            destroy_scope(ctx->fwd_scope);
+            ctx->fwd_scope = NULL;
+        }
+        if (ctx->back_scope) {
+            destroy_scope(ctx->back_scope);
+            ctx->back_scope = NULL;
+        }
+        if (ctx->current_looptree) {
+            destroy_loop_tree(ctx->current_looptree);
+            ctx->current_looptree = NULL;
+        }
+    }
+
+    const Node* old_abstraction = ctx->current_abstraction;
+    if (is_abstraction(node)) {
+        ctx->current_abstraction = node;
+    }
+    const Node* result;
+
+    if (node->tag == Branch_TAG) {
+        if (!ctx->fwd_scope) {
+            ctx->fwd_scope = new_scope(ctx->current_fn);
+            ctx->back_scope = new_scope_flipped(ctx->current_fn);
+            ctx->current_looptree = build_loop_tree(ctx->fwd_scope);
+        }
+
+        CFNode* cfnode = scope_lookup(ctx->back_scope, ctx->current_abstraction);
+
+        CFNode* idom = NULL;
+
+        LTNode* current_loop = looptree_lookup(ctx->current_looptree, ctx->current_abstraction)->parent;
+        assert(current_loop);
+
+        if (entries_count_list(current_loop->cf_nodes)) {
+            bool leaves_loop = false;
+            CFNode* current_node = scope_lookup(ctx->fwd_scope, ctx->current_abstraction);
+            for (size_t i = 0; i < entries_count_list(current_node->succ_edges); i++) {
+                CFEdge edge = read_list(CFEdge, current_node->succ_edges)[i];
+                LTNode* lt_target = looptree_lookup(ctx->current_looptree, edge.dst->node);
+
+                if (lt_target->parent != current_loop) {
```

Comment:
is this correct ? the target could be in an inner loop that is still a children of the parent, transitively